Selection fails to update during autoscrolling as well as keyboard scrolling.

megangardner opened the pull request:
#### c79a81698e5532e5f3a8de9720e3f640a674e683
<pre>
Selection fails to update during autoscrolling as well as keyboard scrolling.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287454">https://bugs.webkit.org/show_bug.cgi?id=287454</a>
<a href="https://rdar.apple.com/144581646">rdar://144581646</a>

Reviewed by Aditya Keerthi and Wenson Hsieh.

When a user uses the keyboard or the autoscroll functionality to
scroll a view while a selection is happening, they don&apos;t generate
additional touches unless their finger moves. Thus if you
hold your finger still while scrolling, the selection doesn&apos;t update.
In order to for a selection to update, we keep track of the position of
the last touch and use it to update the selection when the view is scrolled.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidScroll:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView updateSelection]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):

Canonical link: <a href="https://commits.webkit.org/290299@main">https://commits.webkit.org/290299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a57015d226e4e10596602a0274cac64ce377bac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94557 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40332 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17371 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26639 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7271 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49352 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7002 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96385 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16747 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77857 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77173 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19049 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21604 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9914 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16760 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22074 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16501 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19952 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18283 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5269,12 +5269,46 @@ - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(WKBEGestureType
 
     _autocorrectionContextNeedsUpdate = YES;
     _usingGestureForSelection = YES;
+    auto touchType = toSelectionTouch(touch);
```
```suggestion
    
    auto touchType = toSelectionTouch(touch);
```

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5269,12 +5269,46 @@ - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(WKBEGestureType
+    case WebKit::SelectionTouch::EndedNotMoving:
+        _isInSelectionInteraction = NO;
+        break;
+    }
```
```suggestion
    }
    
```

pxlcoder reviewed (approved):
> during auto

"during auto" is a bit unclear, can you update to say "during autoscroll"?

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5269,12 +5269,46 @@ - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(WKBEGestureType
+    case WebKit::SelectionTouch::Moved:
+        if (_isInSelectionInteraction) {
+            _lastSelectionTouch.point = [self convertPoint:point toView:self.webView];
+            _lastSelectionTouch.baseIsStart = baseIsStart;
```
Would you need to update the last cached `_lastSelectionTouch.baseIsStart` after completing a selection change, in the case where the selection flips? (That way, we wouldn't end up sending `_lastSelectionTouch.baseIsStart` with a stale value of `baseIsStart`)

megangardner replied:
I tested flipping and it was fine, but I’ll double check and also probably do what you suggest anyways.

whsieh replied:
Hm... it probably works fine as long as the user moves (at least slightly) after the selection flips, because then `changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:` will be called with the updated `baseIsStart`. However, if you keep the selection still after flipping, it's unclear how the last cached `baseIsStart` would be updated to reflect the flipping.

megangardner replied:
I think it would be very difficult to flip and then stop moving, so it would be really difficult to get into that state, but I'll fix it :)

megangardner committed:
Selection fails to update during autoscrolling as well as keyboard scrolling.
https://bugs.webkit.org/show_bug.cgi?id=287454
rdar://144581646

Reviewed by Aditya Keerthi and Wenson Hsieh.

When a user uses the keyboard or the autoscroll functionality to
scroll a view while a selection is happening, they don't generate
additional touches unless their finger moves. Thus if you
hold your finger still while scrolling, the selection doesn't update.
In order to for a selection to update, we keep track of the position of
the last touch and use it to update the selection when the view is scrolled.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewDidScroll:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView updateSelection]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):

Canonical link: https://commits.webkit.org/290299@main
